### PR TITLE
fix(language-service): bug of accessing a string index signature using dot notation

### PR DESCRIPTION
**Pull request by ivanwonder**
fix bug of the pr #33884 @ayazhafiz 
```
primitiveType: {[name: string]: string} = {};
```
`{{primitiveType.test}}` this will produce errors. because the `string` type doesn't have Symbol.
https://github.com/angular/angular/blob/168abc6d6f52713383411b14980e104c99bfeef5/packages/language-service/src/typescript_symbols.ts#L509

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/language-service/src/typescript_symbols.ts [hunk 1]
```diff
@@ -308,6 +314,10 @@ class TypeWrapper implements Symbol {
   }
 }
 
+class StringIndexTypeWrappr extends TypeWrapper {
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]:
```suggestion
class StringIndexTypeWrapper extends TypeWrapper {
```

Can you also add documentation for this class like you did on line 522?

File: packages/language-service/src/typescript_symbols.ts [hunk 2]
```diff
@@ -506,10 +516,11 @@ class SymbolTableWrapper implements SymbolTable {
...
       //
       // In this case, return the type indexed by an arbitrary string key.
-      const symbol = this.stringIndexType.getSymbol();
-      if (symbol) {
-        return new SymbolWrapper(symbol, this.context, this.stringIndexType);
-      }
+
+      // if stringIndexType is js primitive type(e.g. 'string'), the Symbol is undefined;
+      // and In AstType.resolvePropertyRead method, the Symbol.type should get the right type.
+      // so I add a new Symbol type, 'StringIndexTypeWrappr'
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]:
```suggestion
```

File: packages/language-service/src/typescript_symbols.ts [hunk 3]
```diff
@@ -506,10 +516,11 @@ class SymbolTableWrapper implements SymbolTable {
       //   obj.stringIndex // equivalent to obj['stringIndex'];
       //
       // In this case, return the type indexed by an arbitrary string key.
-      const symbol = this.stringIndexType.getSymbol();
-      if (symbol) {
-        return new SymbolWrapper(symbol, this.context, this.stringIndexType);
-      }
+
+      // if stringIndexType is js primitive type(e.g. 'string'), the Symbol is undefined;
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]:
```suggestion
      // If stringIndexType a primitive type(e.g. 'string'), the Symbol is undefined;
```

File: packages/language-service/src/typescript_symbols.ts [hunk 4]
```diff
@@ -506,10 +516,11 @@ class SymbolTableWrapper implements SymbolTable {
...
+      // and In AstType.resolvePropertyRead method, the Symbol.type should get the right type.
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]:
```suggestion
      // and in AstType.resolvePropertyRead method, the Symbol.type should get the right type.
```

File: packages/language-service/test/diagnostics_spec.ts [hunk 5]
```diff
@@ -147,6 +147,14 @@ describe('diagnostics', () => {
         expect(diags[0].messageText)
             .toBe(`Identifier 'badProperty' is not defined. 'Hero' does not contain such a member`);
       });
+
+      it('should not produce errors with dot notation if stringIndexType is js primitive type',
+         () => {
+           mockHost.override(TEST_TEMPLATE, `
+        {{primitiveType.test}}`);
```
Comment (ayazhafiz) [inline packages/language-service/test/diagnostics_spec.ts]: Can you place this template on one line? Also, can you add a test in `completions_spec.ts`? There are similar tests in there already.

File: packages/language-service/test/project/app/parsing-cases.ts [hunk 6]
```diff
@@ -192,6 +192,7 @@ export class TemplateReference {
   tupleArray: [string, Hero] = ['test', this.hero];
   league: Hero[][] = [this.heroes];
   heroesByName: {[name: string]: Hero} = {};
+  primitiveType: {[name: string]: string} = {};
```
Comment (ayazhafiz) [inline packages/language-service/test/project/app/parsing-cases.ts]: Can you rename to `primitiveIndexType` or similar? `primitiveType` sounds like the type (of the wrapper) itself is primitive.

**Review by ayazhafiz — CHANGES_REQUESTED** [review]

File: packages/language-service/src/typescript_symbols.ts [hunk 7]
```diff
@@ -246,7 +246,13 @@ class TypeWrapper implements Symbol {
 
   get name(): string {
     const symbol = this.tsType.symbol;
-    return (symbol && symbol.name) || '<anonymous>';
+    if (symbol) {
+      return symbol.name;
+    } else {
+      // the js primitive type(e.g. 'string') doesn't have Symbol.
```
[the remark this answers is not in the document]
Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: yes, you are right.
for example
```
tupleArray: [string, Hero];
```
this will return `[string, Hero]`

File: packages/language-service/src/typescript_symbols.ts (hunk 1 above)
Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Can you fix the typo here?

File: packages/language-service/src/typescript_symbols.ts [hunk 8]
```diff
@@ -249,7 +249,7 @@ class TypeWrapper implements Symbol {
     if (symbol) {
       return symbol.name;
     } else {
-      // the js primitive type(e.g. 'string') doesn't have Symbol.
+      // a primitive type (e.g. 'string') doesn't have Symbol.
       // use the ts.TypeChecker to get the type name.
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]: Could you capitalize the sentences here?

File: packages/language-service/src/typescript_symbols.ts [hunk 9]
```diff
@@ -517,9 +519,8 @@ class SymbolTableWrapper implements SymbolTable {
       //
       // In this case, return the type indexed by an arbitrary string key.
 
-      // if stringIndexType is js primitive type(e.g. 'string'), the Symbol is undefined;
-      // and In AstType.resolvePropertyRead method, the Symbol.type should get the right type.
-      // so I add a new Symbol type, 'StringIndexTypeWrappr'
+      // If stringIndexType a primitive type(e.g. 'string'), the Symbol is undefined;
+      // and in AstType.resolvePropertyRead method, the Symbol.type should get the right type.
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]: I think you can get rid of the comment here, since you added them on the new class :)

**Review by ayazhafiz — CHANGES_REQUESTED** [review]
LGTM, just a couple of typos.

**Review by ayazhafiz — APPROVED** [review]

**Comment by ayazhafiz** [comment]
Thank you!

File: packages/language-service/src/typescript_symbols.ts [hunk 10]
```diff
@@ -246,7 +246,13 @@ class TypeWrapper implements Symbol {
 
   get name(): string {
     const symbol = this.tsType.symbol;
-    return (symbol && symbol.name) || '<anonymous>';
+    if (symbol) {
+      return symbol.name;
```
Comment (kyliau) [inline packages/language-service/src/typescript_symbols.ts]: Do we still need to check the symbol here?
I think the else case here (getting name from the typechecker) is more generic and will always return the right name.

Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: I think it should be removed, but this code uses it as symbol name.
https://github.com/angular/angular/blob/37fa0837954865d3653bcdc4564e732fbf8b8691/packages/language-service/src/typescript_symbols.ts#L620
for example, the typechecker will return `Promise<string>`, the symbol name is `Promise`. It confuses me. or I change it to `parameterType.tsType.symbol.name`.

Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Checking for symbol can be removed by changing the code in `PipeSymbol#selectSignature` from the current evaluation of the piped object's type parameter:

https://github.com/angular/angular/blob/37fa0837954865d3653bcdc4564e732fbf8b8691/packages/language-service/src/typescript_symbols.ts#L619-L630

to an evaluation of the type parameter using the typechecker:

```typescript
case 'async':
  const typeNode = this.context.checker.typeToTypeNode(parameterType.tsType);
  if (typeNode && ts.isTypeReferenceNode(typeNode) && 
      ts.isIdentifier(typeNode.typeName)) {
    const name = typeNode.typeName.text;
    const result = getTypeParameterOf(parameterType.tsType, name);
    if (result) resultType = result;
  }
  break;
```

(Maybe it's worth also adding a default of `any` to stay consistent with the current implementation, but its exclusion doesn't break any tests.)

Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: I test it before, maybe the `any` has something to do with this test:
https://github.com/angular/angular/blob/37fa0837954865d3653bcdc4564e732fbf8b8691/packages/language-service/test/diagnostics_spec.ts#L568

Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Can you elaborate? I didn’t have a problem with `any` with the implementation I commented.

Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: https://github.com/angular/angular/blob/37fa0837954865d3653bcdc4564e732fbf8b8691/packages/language-service/src/typescript_symbols.ts#L620
In the test of 'parse a template reference', the `parameterType` type is `any`, then the code will go to line 627.
Your code works fine. I mean the code before applying your modifications.

Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Ah, my bad; thank you for explaining. Either way, exclusion of the branch on line 627 should be okay since an `any` type will match any signature.

Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]:
```
switch (this.name) {
          case 'async':
            const symbol = parameterType.tsType.symbol;
            if (symbol) {
              resultType = getTypeParameterOf(parameterType.tsType, symbol.name);
            }
            break;
          case 'slice':
            resultType = parameterType.tsType;
            break;
        }
```
why not use `parameterType.tsType.symbol` just I have said, it's simpler and easier to read

Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Yes, it is much simpler, and I think it would be great to use it in the way you suggest; the only reason I suggested the other one is because it doesn't rely on the bad-typed `symbol` property and is symmetric to your new implementation of `get name()`.

File: packages/language-service/src/typescript_symbols.ts [hunk 11]
```diff
@@ -608,15 +608,10 @@ class PipeSymbol implements Symbol {
...
-            switch (parameterType.name) {
-              case 'Observable':
-              case 'Promise':
-              case 'EventEmitter':
-                resultType = getTypeParameterOf(parameterType.tsType, parameterType.name);
-                break;
-              default:
-                resultType = getTsTypeFromBuiltinType(BuiltinType.Any, this.context);
-                break;
+            // It's safe to use Symbol here for the type 'Observable, Promise, EventEmitter'.
```
Comment (ayazhafiz) [inline packages/language-service/src/typescript_symbols.ts]: Could you rephrase or elaborate this comment? What do you mean by safe?
```suggestion
            // It's safe to use Symbol here for the type 'Observable', 'Promise', or 'EventEmitter'.
```

**Review by ayazhafiz — APPROVED** [review]

File: packages/language-service/src/typescript_symbols.ts (hunk 11 above)
Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: this place only needs to handle these three types. It's enough to use Symbol. Maybe replace it with 'enough'.

Comment (ayazhafiz) [reply packages/language-service/src/typescript_symbols.ts]: Consider rephrasing to `Get symbol of 'Observable', 'Promise', or 'EventEmitter' type` or similar.

Comment (ivanwonder) [reply packages/language-service/src/typescript_symbols.ts]: How about `Get TypeParameter of 'Observable', 'Promise', or 'EventEmitter' type`

**Review by kyliau — APPROVED** [review]
Thank you very much @ivanwonder
This solution is clean and elegant, much appreciated!
